secrethub_wrapper.go

```diff
@@ -0,0 +1,187 @@
+			Name:    "secrethub-xgo",
+			Version: "0.1.0",
+		}),
+	}
+	client, err := secrethub.NewClient(options...)
+	return client, err
+}
+
+// Read retrieves a secret by its path.
+//export Read
+func Read(path *C.char, errMessage **C.char) C.struct_SecretVersion {
+	client, err := Client()
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return C.struct_SecretVersion{}
+	}
+	secret, err := client.Secrets().Read(C.GoString(path))
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return C.struct_SecretVersion{}
+	}
+	return C.struct_SecretVersion{
+		SecretVersionID: C.CString(secret.SecretVersionID.String()),
+		Secret: C.struct_Secret{
+			SecretID:      C.CString(secret.Secret.SecretID.String()),
+			DirID:         C.CString(secret.Secret.DirID.String()),
+			RepoID:        C.CString(secret.Secret.RepoID.String()),
+			Name:          C.CString(secret.Secret.Name),
+			BlindName:     C.CString(secret.Secret.BlindName),
+			VersionCount:  C.int(secret.Secret.VersionCount),
+			LatestVersion: C.int(secret.Secret.LatestVersion),
+			Status:        C.CString(secret.Secret.Status),
+			CreatedAt:     C.longlong(secret.Secret.CreatedAt.Unix()),
+		},
+		Version:   C.int(secret.Version),
+		Data:      C.CString(string(secret.Data)),
+		CreatedAt: C.longlong(secret.CreatedAt.Unix()),
+		Status:    C.CString(secret.Status),
+	}
+}
+
+// ReadString retrieves a secret as a string.
+//export ReadString
+func ReadString(path *C.char, errMessage **C.char) *C.char {
+	client, err := Client()
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
+	secret, err := client.Secrets().Read(C.GoString(path))
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
+	return C.CString(string(secret.Data))
+}
+
+// Resolve fetches the values of a secret from SecretHub, when the `ref` parameter
+// has the format `secrethub://<path>`. Otherwise it returns `ref` unchanged, as an array of bytes.
+//export Resolve
+func Resolve(ref *C.char, errMessage **C.char) *C.char {
+	client, err := Client()
+	if err != nil {
+		*errMessage = C.CString(err.Error())
+		return nil
+	}
+	bits := strings.Split(C.GoString(ref), "://")
+	if len(bits) == 2 && bits[0] == "secrethub" {
+		secret, err := client.Secrets().Read(bits[1])
+		if err != nil {
+			*errMessage = C.CString(err.Error())
+			return nil
+		}
+		return C.CString(string(secret.Data))
+	}
+	return ref
+}
+
+// ResolveEnv takes a map of environment variables and replaces the values of those
+// which store references of secrets in SecretHub (`secrethub://<path>`) with the value
+// of the respective secret. The other entries in the map remain untouched.
+//export ResolveEnv
+func ResolveEnv(errMessage **C.char) *C.char {
+	envVars := os.Environ()
+	resolvedEnv := make(map[string]string, len(envVars))
+	for _, value := range envVars {
+		envVar := strings.Split(value, "=")
+		key := envVar[0]
+		value := C.CString(envVar[1])
```

Comment:
According to the documentation, [os.Environ()](https://golang.org/pkg/os/#Environ) returns in `key=value` format. Therefore, there should be no concerns about lines without an `=` in them.
However, in the adjustments that I will make, we will ignore all lines where the length ov `envVar` is 1, even though this will never happen.

Reply:
Hmm, you're right. 

I was confused by the weird entry that Windows sometimes adds to the `Environ()` list and that has previously bitten up (because it's really weird): `:=::=`. But entry still adheres to the rule that every entry contains a `=`. 

Nevertheless, the `== 1` check does no harm, so we can still leave it in.

Reply:
The condition is there.